lib/hanami/mailer.rb

```diff
@@ -38,300 +25,265 @@ def initialize
       txt:  'text/plain'
     }.freeze
 
-    include Utils::ClassAttribute
+    private_constant(:CONTENT_TYPES)
 
+    # Base error for Hanami::Mailer
+    #
     # @since 0.1.0
-    # @api private
-    class_attribute :configuration
-    self.configuration = Configuration.new
+    class Error < ::StandardError
+    end
 
-    # Configure the framework.
-    # It yields the given block in the context of the configuration
+    # Unknown mailer
     #
-    # @param blk [Proc] the configuration block
+    # This error is raised at the runtime when trying to deliver a mail message,
+    # by using a configuration that it wasn't finalized yet.
     #
-    # @since 0.1.0
+    # @since next
+    # @api unstable
     #
-    # @see Hanami::Mailer::Configuration
+    # @see Hanami::Mailer.finalize
+    class UnknownMailerError < Error
+      # @param mailer [Hanami::Mailer] a mailer
+      #
+      # @since next
+      # @api unstable
+      def initialize(mailer)
+        super("Unknown mailer: #{mailer.inspect}. Please finalize the configuration before to use it.")
+      end
+    end
+
+    # Missing delivery data error
     #
-    # @example
-    #   require 'hanami/mailer'
+    # It's raised when a mailer doesn't specify `from` or `to`.
     #
-    #   Hanami::Mailer.configure do
-    #     root '/path/to/root'
-    #   end
-    def self.configure(&blk)
-      configuration.instance_eval(&blk)
-      self
+    # @since 0.1.0
+    class MissingDeliveryDataError < Error
+      def initialize
+        super("Missing delivery data, please check 'from', or 'to'")
+      end
     end
 
+    # @since next
+    # @api unstable
+    @_subclasses = Concurrent::Array.new
+
     # Override Ruby's hook for modules.
-    # It includes basic Hanami::Mailer modules to the given Class.
+    # It includes basic `Hanami::Mailer` modules to the given Class.
     # It sets a copy of the framework configuration
     #
     # @param base [Class] the target mailer
     #
-    # @since 0.1.0
-    # @api private
-    #
-    # @see http://www.ruby-doc.org/core/Module.html#method-i-included
-    def self.included(base)
-      conf = configuration
-      conf.add_mailer(base)
-
-      base.class_eval do
-        extend Dsl
-        extend ClassMethods
-
-        include Utils::ClassAttribute
-        class_attribute :configuration
-
-        self.configuration = conf.duplicate
-      end
-
-      conf.copy!(base)
+    # @since next
+    # @api unstable
+    def self.inherited(base)
+      @_subclasses.push(base)
+      base.extend Dsl
```

Comment:
Still trying to wrap my head around the changes, but can't we just extend `Hanami::Mailer` and not subvert the hook so it adds methods not defined on the superclass?